core/src/main/java/bio/terra/pearl/core/service/participant/merge/MergeAction.java

```diff
@@ -0,0 +1,51 @@
+package bio.terra.pearl.core.service.participant.merge;
+
+import bio.terra.pearl.core.model.BaseEntity;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.Setter;
+
+@Builder
+@Getter
+@Setter
+public class MergeAction<T extends BaseEntity, MP> {
+    private MergePair<T> pair;
+    private Action action;
+    private MP mergePlan; // if the action is MERGE, this is the merge plan
+
+    public MergeAction(MergePair<T> pair, Action action) {
+        this.pair = pair;
+        this.action = action;
+    }
+
+    public MergeAction(MergePair<T> pair, Action action, MP mergePlan) {
+        if (action ==Action.MOVE_SOURCE && pair.getSource() == null) {
+            throw new IllegalArgumentException("source cannot be null for MOVE_SOURCE action");
+        }
+        if (action ==Action.DELETE_SOURCE && pair.getSource() == null) {
+            throw new IllegalArgumentException("source cannot be null for DELETE_SOURCE action");
+        }
+        if (action == Action.MERGE && (pair.getSource() == null || pair.getTarget() == null)) {
+            throw new IllegalArgumentException("MERGE requires a source, target and plan");
+        }
+        this.pair = pair;
+        this.action = action;
+        this.mergePlan = mergePlan;
+    }
+
+    public T getSource() {
+        return pair.getSource();
+    }
+
+    public T getTarget() {
+        return pair.getTarget();
+    }
+
+    public enum Action {
+        MOVE_SOURCE,
+        NO_ACTION,
+        MERGE,
+        DELETE_SOURCE,
+        MOVE_SOURCE_DELETE_TARGET
```

Comment:
It took me a while to fully grok the approach here. I would find it to be super helpful to have a basic `README.md` explaining the overall merge approach and also some comments here to explain the purpose of each `Action` (e.g., MOVE_SOURCE_DELETE_TARGET is for when target has no data but source does, etc). I think there's a lot of intentional scoping here to make it not too complex, but it's not really clear what that scoping is, so if I had to touch this code I think I would have to ask a lot of questions without just a bit more documentation